File: 347-top-k-frequent-elements/top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const freq = new Map()
    for (let num of nums) {
        freq.set(num, (freq.get(num) || 0) + 1)
    }

    const maxHeap = new MaxPriorityQueue()
    for (let [num, count] of freq) {
        maxHeap.enqueue(num, count)
    }

    const result = []
    for (let i = 0; i < k; i++) {
        result.push(maxHeap.dequeue().element)
    }

    return result
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = Counter(nums)

        buckets = [[] for _ in range(len(nums) + 1)]
        for num, count in freq.items():
            buckets[count].append(num)

        res = []

        for count in range(len(buckets) - 1, 0, -1):
            for num in buckets[count]:
                res.append(num)
                if len(res) == k:
                    return res
        
        return res